engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java

```diff
@@ -428,8 +438,31 @@ private DirectBuffer insertAt(
 
     if (completionCondition.isPresent()) {
       return expressionBehavior.evaluateBooleanExpression(
-          completionCondition.get(), context.getElementInstanceKey());
+          completionCondition.get(),
+          context.getElementInstanceKey(),
+          (variableName) -> getVariable(element, context, variableName));
     }
     return Either.right(false);
   }
+
+  private DirectBuffer getVariable(
+      final ExecutableMultiInstanceBody element,
+      final BpmnElementContext context,
+      final String variableName) {
+    return switch (variableName) {
+      case "numberOfInstances" -> wrapVariable(
+          numberOfInstancesVariableBuffer,
+          numberOfInstancesVariableView,
+          readInputCollectionVariable(element, context).get().size());
```

Comment:
I think the second option is probably the best option. 🤔

Reply:
Hi @Lzgabel I was rereading the spec on these properties again

> [about `numberOfInstances`] This attribute is provided for the outer instance of the Multi-Instance Activity only. This attribute contains the total number of inner instances created for the Multi-Instance Activity

So it is not the size of the input collection, but rather the number of created instances. 

This is further acknowledged by:

> [about `numberOfTerminatedInstances`] The sum of numberOfTerminatedInstances, numberOfCompletedInstances, and numberOfActiveInstances always sums up to numberOfInstances.

So in a sequential multi-instance with an input collection size of 3 and they all complete instead of terminate, the values count like this:
```
numberOfInstances 1, numberOfActiveInstances 1, numberOfCompletedInstances 0, numberOfTerminatedInstances 0
numberOfInstances 2, numberOfActiveInstances 1, numberOfCompletedInstances 1, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 1, numberOfCompletedInstances 2, numberOfTerminatedInstances 0
// and once the last one is completed
numberOfInstances 3, numberOfActiveInstances 0, numberOfCompletedInstances 3, numberOfTerminatedInstances 0
```

For a parallel multi-instance:
```
numberOfInstances 3, numberOfActiveInstances 3, numberOfCompletedInstances 0, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 2, numberOfCompletedInstances 1, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 1, numberOfCompletedInstances 2, numberOfTerminatedInstances 0
numberOfInstances 3, numberOfActiveInstances 0, numberOfCompletedInstances 3, numberOfTerminatedInstances 0
```